Session-scoped fixture executed too many times with `-x` / `--maxfails`

This is related, but I think slightly different than #868 and #420 (which deals with the fact that these flags do not stop execution after a fail).

Observe this test file:
```python
import sys
from time import sleep

from pytest import fixture

# hack for live prints (https://stackoverflow.com/questions/27006884)
sys.stdout = sys.stderr


@fixture(scope='session', autouse=True)
def session_fixture():
    print('### executing fixture! ###')


def test_1():
    assert False


def test_2():
    sleep(1)


def test_3():
    sleep(1)


def test_4():
    sleep(1)
```

Running this with: 
```
pytest test_parallel.py  -n 2  -s --maxfail=1
```

leads to the `session_fixture` being executed 3 times!   Removing `--maxfail=1`, you can see `session_fixture` is executed twice (as expected).

I did some more investigation into this - I don't think the issue is in `pytest`, but in the way `pytest-xdist` is (mis)handling the `--maxfail` flag. For example, if you run the example code above using `pytest` version 7.4.4, multiple tests will run _on a single worker_ before the worker quits. I think this is incorrect...after a single failure on a worker, it should exit.

A simpler way to see this is using `-n 1`. As I understand it, `-n 1` should behave the same as simply omitting `-n`.

```
% pytest test_parallel.py  -n 1 -s --maxfail=1
====================================================== test session starts =======================================================
platform darwin -- Python 3.8.13, pytest-7.4.4, pluggy-1.4.0
rootdir: /Users/bbrown/Downloads
plugins: xdist-3.5.0
1 worker [4 items]
### executing fixture! ###
F.
============================================================ FAILURES ============================================================
_____________________________________________________________ test_1 _____________________________________________________________
[gw0] darwin -- Python 3.8.13 /Users/bbrown/Downloads/venv_old_pytest/bin/python

    def test_1():
>       assert False
E       assert False

test_parallel.py:16: AssertionError
==================================================== short test summary info =====================================================
FAILED test_parallel.py::test_1 - assert False
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! xdist.dsession.Interrupted: stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================== 1 failed, 1 passed in 1.41s ===================================================
```

With `-n 2`, you can still see that `gw0` runs a second test after the first one fails. I think `pytest` version 8 just exposes this issue, since the hooks in `pytest-xdist` keep running when they should have exited.

I have a fix locally, I'll try to get into a pull request.

--- a/changelog/1024.bugfix
+++ b/changelog/1024.bugfix
@@ -1,0 +1,2 @@
+Added proper handling of ``shouldstop`` (such as set by ``--max-fail``) and ``shouldfail`` conditions in workers.
+Previously, a worker might have continued executing further tests before the controller could terminate the session.

--- a/src/xdist/dsession.py
+++ b/src/xdist/dsession.py
@@ -182,18 +182,17 @@ class DSession:
             self.shouldstop = f"{node} received keyboard-interrupt"
             self.worker_errordown(node, "keyboard-interrupt")
             return
-
         shouldfail = node.workeroutput["shouldfail"]
         shouldstop = node.workeroutput["shouldstop"]
         for shouldx in [shouldfail, shouldstop]:
             if shouldx:
-                self.shouldstop = shouldx
-                self._active_nodes.remove(node)
-                return
-
-        if node in self.sched.nodes:
-            crashitem = self.sched.remove_node(node)
-            assert not crashitem, (crashitem, node)
+                if not self.shouldstop:
+                    self.shouldstop = shouldx
+                break
+        else:
+            if node in self.sched.nodes:
+                crashitem = self.sched.remove_node(node)
+                assert not crashitem, (crashitem, node)
         self._active_nodes.remove(node)
 
     def worker_internal_error(self, node, formatted_error):

--- a/src/xdist/dsession.py
+++ b/src/xdist/dsession.py
@@ -182,7 +182,7 @@ class DSession:
             self.shouldstop = f"{node} received keyboard-interrupt"
             self.worker_errordown(node, "keyboard-interrupt")
             return
-        if not self.shouldstop and node in self.sched.nodes:
+        if node in self.sched.nodes:
             crashitem = self.sched.remove_node(node)
             assert not crashitem, (crashitem, node)
         self._active_nodes.remove(node)

--- a/src/xdist/dsession.py
+++ b/src/xdist/dsession.py
@@ -182,7 +182,16 @@ class DSession:
             self.shouldstop = f"{node} received keyboard-interrupt"
             self.worker_errordown(node, "keyboard-interrupt")
             return
-        if not self.shouldstop and node in self.sched.nodes:
+
+        shouldfail = node.workeroutput["shouldfail"]
+        shouldstop = node.workeroutput["shouldstop"]
+        for shouldx in [shouldfail, shouldstop]:
+            if shouldx:
+                self.shouldstop = shouldx
+                self._active_nodes.remove(node)
+                return
+
+        if node in self.sched.nodes:
             crashitem = self.sched.remove_node(node)
             assert not crashitem, (crashitem, node)
         self._active_nodes.remove(node)

--- a/src/xdist/dsession.py
+++ b/src/xdist/dsession.py
@@ -182,7 +182,7 @@ class DSession:
             self.shouldstop = f"{node} received keyboard-interrupt"
             self.worker_errordown(node, "keyboard-interrupt")
             return
-        if node in self.sched.nodes:
+        if not self.shouldstop and node in self.sched.nodes:
             crashitem = self.sched.remove_node(node)
             assert not crashitem, (crashitem, node)
         self._active_nodes.remove(node)

--- a/src/xdist/remote.py
+++ b/src/xdist/remote.py
@@ -155,6 +155,8 @@ class WorkerInteractor:
         self.nextitem_index = self._get_next_item_index()
         while self.nextitem_index is not self.SHUTDOWN_MARK:
             self.run_one_test()
+            if session.shouldfail or session.shouldstop:
+                break
         return True
 
     def run_one_test(self):

--- a/src/xdist/remote.py
+++ b/src/xdist/remote.py
@@ -154,12 +154,10 @@ class WorkerInteractor:
         self.channel.setcallback(self.handle_command, endmarker=self.SHUTDOWN_MARK)
         self.nextitem_index = self._get_next_item_index()
         while self.nextitem_index is not self.SHUTDOWN_MARK:
-            self.run_one_test()
-            if session.shouldfail or session.shouldstop:
-                break
+            self.run_one_test(session.shouldfail or session.shouldstop)
         return True
 
-    def run_one_test(self):
+    def run_one_test(self, skip: bool):
         self.item_index = self.nextitem_index
         self.nextitem_index = self._get_next_item_index()
 
@@ -173,7 +171,10 @@ class WorkerInteractor:
         worker_title("[pytest-xdist running] %s" % item.nodeid)
 
         start = time.time()
-        self.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
+
+        if not skip:
+            self.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
+
         duration = time.time() - start
 
         worker_title("[pytest-xdist idle]")

--- a/src/xdist/remote.py
+++ b/src/xdist/remote.py
@@ -154,10 +154,12 @@ class WorkerInteractor:
         self.channel.setcallback(self.handle_command, endmarker=self.SHUTDOWN_MARK)
         self.nextitem_index = self._get_next_item_index()
         while self.nextitem_index is not self.SHUTDOWN_MARK:
-            self.run_one_test(session.shouldfail or session.shouldstop)
+            self.run_one_test()
+            if session.shouldfail or session.shouldstop:
+                break
         return True
 
-    def run_one_test(self, skip: bool):
+    def run_one_test(self):
         self.item_index = self.nextitem_index
         self.nextitem_index = self._get_next_item_index()
 
@@ -171,10 +173,7 @@ class WorkerInteractor:
         worker_title("[pytest-xdist running] %s" % item.nodeid)
 
         start = time.time()
-
-        if not skip:
-            self.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
-
+        self.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
         duration = time.time() - start
 
         worker_title("[pytest-xdist idle]")

--- a/src/xdist/remote.py
+++ b/src/xdist/remote.py
@@ -103,6 +103,8 @@ class WorkerInteractor:
     def pytest_sessionfinish(self, exitstatus):
         # in pytest 5.0+, exitstatus is an IntEnum object
         self.config.workeroutput["exitstatus"] = int(exitstatus)
+        self.config.workeroutput["shouldfail"] = self.session.shouldfail
+        self.config.workeroutput["shouldstop"] = self.session.shouldstop
         yield
         self.sendevent("workerfinished", workeroutput=self.config.workeroutput)
 

--- a/testing/acceptance_test.py
+++ b/testing/acceptance_test.py
@@ -143,11 +143,9 @@ class TestDistribution:
         assert result.ret == 2
         # DSession should stop when the first failure is reached. Two failures
         # may actually occur, due to timing.
-        result.stdout.re_match_lines([".*Interrupted: stopping.*[12].*"])
-        m = re.search(r"== (\d+) failed", str(result.stdout))
-        assert m
-        n_failed = int(m.groups()[0])
-        assert 1 <= n_failed <= 2
+        outcomes = result.parseoutcomes()
+        assert "failed" in outcomes, "Expected at least one failure"
+        assert 1 <= outcomes["failed"] <= 2, "Expected no more than 2 failures"
 
     def test_basetemp_in_subprocesses(self, pytester: pytest.Pytester) -> None:
         p1 = pytester.makepyfile(
@@ -1195,7 +1193,7 @@ def test_maxfail_causes_early_termination(pytester: pytest.Pytester) -> None:
     """
     )
     result = pytester.runpytest_subprocess("--maxfail=1", "-n 1")
-    result.stdout.re_match_lines([".* 1 failed in .*"])
+    result.assert_outcomes(failed=1)
 
 
 def test_internal_errors_propagate_to_controller(pytester: pytest.Pytester) -> None:

--- a/testing/acceptance_test.py
+++ b/testing/acceptance_test.py
@@ -109,12 +109,12 @@ class TestDistribution:
         )
         assert result.ret == 1
 
-    def test_exitfail_waits_for_workers_to_finish(
+    def test_exitfirst_waits_for_workers_to_finish(
         self, pytester: pytest.Pytester
     ) -> None:
         """The DSession waits for workers before exiting early on failure.
 
-        When -x/--exitfail is set, the DSession wait for the workers to finish
+        When -x/--exitfirst is set, the DSession wait for all workers to finish
         before raising an Interrupt exception. This prevents reports from the
         faiing test and other tests from being discarded.
         """
@@ -138,15 +138,16 @@ class TestDistribution:
                 time.sleep(0.3)
         """
         )
+        # Two workers are used
         result = pytester.runpytest(p1, "-x", "-rA", "-v", "-n2")
         assert result.ret == 2
+        # DSession should stop when the first failure is reached. Two failures
+        # may actually occur, due to timing.
         result.stdout.re_match_lines([".*Interrupted: stopping.*[12].*"])
-        m = re.search(r"== (\d+) failed, (\d+) passed in ", str(result.stdout))
+        m = re.search(r"== (\d+) failed", str(result.stdout))
         assert m
-        n_failed, n_passed = (int(s) for s in m.groups())
+        n_failed = int(m.groups()[0])
         assert 1 <= n_failed <= 2
-        assert 1 <= n_passed <= 3
-        assert (n_passed + n_failed) < 6
 
     def test_basetemp_in_subprocesses(self, pytester: pytest.Pytester) -> None:
         p1 = pytester.makepyfile(
@@ -1178,6 +1179,23 @@ def test_internal_error_with_maxfail(pytester: pytest.Pytester) -> None:
     result = pytester.runpytest_subprocess("--maxfail=1", "-n1")
     result.stdout.re_match_lines([".* [12] errors? in .*"])
     assert "INTERNALERROR" not in result.stderr.str()
+
+
+def test_maxfail_causes_early_termination(pytester: pytest.Pytester) -> None:
+    """
+    Ensure subsequent tests on a worker aren't run when using --maxfail (#1024).
+    """
+    pytester.makepyfile(
+        """
+        def test1():
+            assert False
+
+        def test2():
+            pass
+    """
+    )
+    result = pytester.runpytest_subprocess("--maxfail=1", "-n 1")
+    result.stdout.re_match_lines([".* 1 failed in .*"])
 
 
 def test_internal_errors_propagate_to_controller(pytester: pytest.Pytester) -> None:

